### ISSUE TITLE
baremetal buildkite agent networking / instability issues

### ISSUE THREAD

**Issue by slimsag** [sourcegraph/sourcegraph#12996]
Tracking issue for some networking issues we are seeing with the baremetal buildkite agents

I have a vagrant-based test that is very reliable/fast on my mac, taking about ~10m to complete, but on the baremetal agents network failures plague it for some reason and I haven't been able to get a successful build due to it.

Examples of failures:

Docker pull failures: https://buildkite.com/sourcegraph/deploy-sourcegraph-docker/builds/848#2bec0f33-aa4a-473d-a22d-405b50e3d84c

```
pure-docker-test: c60153a13634: Downloading
--
  | pure-docker-test: c511caab831b: Downloading
  | pure-docker-test: 95e18973d858: Retrying in 15 seconds
  | pure-docker-test: 95e18973d858:
  | pure-docker-test: Retrying in 14 seconds
  | pure-docker-test: 95e18973d858: Retrying in 10 seconds
  | pure-docker-test: 95e18973d858: Retrying in 9 seconds
  | pure-docker-test: docker: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers).
  | pure-docker-test: See 'docker run --help'.
  | pure-docker-test: 95e18973d858: Retrying in 8 seconds
  | pure-docker-test: 95e18973d858: Retrying in 7 seconds
  | pure-docker-test: 95e18973d858: Retrying in 5 seconds
  | pure-docker-test: 15faeb27ba70: Downloading
  | pure-docker-test: 15faeb27ba70: Retrying in 5 seconds
  | The SSH command responded with a non-zero exit status. Vagrant
  | assumes that this means the command failed. The output for this command
  | should be in the log above. Please read the output to determine what
  | went wrong.
  | 🚨 Error: The command exited with status 1
```

SSH failures: https://buildkite.com/sourcegraph/deploy-sourcegraph-docker/builds/847#b72bd8ce-6c87-49c6-980f-b38f00b47ecd

```
==> pure-docker-test: Waiting for machine to boot. This may take a few minutes...
--
  | pure-docker-test: SSH address: 127.0.0.1:2201
  | pure-docker-test: SSH username: vagrant
  | pure-docker-test: SSH auth method: private key
  | pure-docker-test: Warning: Connection reset. Retrying...
  | pure-docker-test: Warning: Connection reset. Retrying...
  | pure-docker-test:
  | pure-docker-test: Vagrant insecure key detected. Vagrant will automatically replace
  | pure-docker-test: this with a newly generated keypair for better security.
  | pure-docker-test:
  | pure-docker-test: Inserting generated public key within guest...
  | Timed out while waiting for the machine to boot. This means that
  | Vagrant was unable to communicate with the guest machine within
  | the configured ("config.vm.boot_timeout" value) time period.
  |  
  | If you look above, you should be able to see the error(s) that
  | Vagrant had when attempting to connect to the machine. These errors
  | are usually good hints as to what may be wrong.
  |  
  | If you're using a custom box, make sure that networking is properly
  | working and you're able to connect to the machine. It is a common
  | problem that networking isn't setup properly in these boxes.
  | Verify that authentication configurations are also setup properly,
  | as well.
```

Apt failures: https://buildkite.com/sourcegraph/deploy-sourcegraph-docker/builds/845#285d3056-5a09-409c-b414-60a61639bb70

```
pure-docker-test: Created symlink /etc/systemd/system/sockets.target.wants/docker.socket → /lib/systemd/system/docker.socket.
--
  | pure-docker-test: Job for docker.service failed because the control process exited with error code.
  | pure-docker-test: See "systemctl status docker.service" and "journalctl -xe" for details.
  | pure-docker-test: invoke-rc.d: initscript docker, action "start" failed.
  | pure-docker-test: ● docker.service - Docker Application Container Engine
  | pure-docker-test:    Loaded: loaded (/lib/systemd/system/docker.service; enabled; vendor preset: enabled)
  | pure-docker-test:    Active: activating (auto-restart) (Result: exit-code) since Thu 2020-08-13 02:18:28 UTC; 26ms ago
  | pure-docker-test:      Docs: https://docs.docker.com
  | pure-docker-test:   Process: 3699 ExecStart=/usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock (code=exited, status=1/FAILURE)
  | pure-docker-test:  Main PID: 3699 (code=exited, status=1/FAILURE)
  | pure-docker-test: dpkg: error processing package docker-ce (--configure):
  | pure-docker-test:  installed docker-ce package post-installation script subprocess returned error exit status 1
  | pure-docker-test: Processing triggers for libc-bin (2.27-3ubuntu1.2) ...
  | pure-docker-test: Processing triggers for systemd (237-3ubuntu10.41) ...
  | pure-docker-test: Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
  | pure-docker-test: Processing triggers for ureadahead (0.100.0-21) ...
  | pure-docker-test: Errors were encountered while processing:
  | pure-docker-test:  docker-ce
  | pure-docker-test: E
  | pure-docker-test: :
  | pure-docker-test: Sub-process /usr/bin/dpkg returned an error code (1)
  | The SSH command responded with a non-zero exit status. Vagrant
  | assumes that this means the command failed. The output for this command
  | should be in the log above. Please read the output to determine what
  | went wrong.
  | 🚨 Error: The command exited with status 1
```

**Comment by slimsag** [comment]
@davejrt I merged my PR into master just now (with some other fixes), you can test by just uncommenting the following: https://github.com/sourcegraph/deploy-sourcegraph-docker/blob/master/.buildkite/pipeline.yaml#L7-L14

**Comment by slimsag** [comment]
I have since been able to reproduce this a few times on Mac + VirtualBox + Vagrant and am worried this may be something to do with VirtualBox<->Ubuntu 18.04

**Comment by davejrt** [comment]
@slimsag I have not been able to get any significant improvement despite best efforts to tune the GCP instance and change vbox network settings or the ssh config for vagrant. It was equally slow when I ran it on windows as well. I'll for arguments sake try with some debian and other ubuntu flavors. 

I think we might consider going with docker and working out how we can create clean environments. 

Unfortunately autoscaling groups in GCP don't function the same as AWS, where you can scale up or down on different metrics, nor can you easily scale to zero. There are some work arounds that involve writing cloud functions to check the metrics then create things on the fly, but that's more code we have to maintain.

**Comment by pecigonzalo** [comment]
@davejrt I believe you should be able to achieve autoscaling with https://cloud.google.com/compute/docs/autoscaler/scaling-stackdriver-monitoring-metrics.

**Comment by davejrt** [comment]
@pecigonzalo we have autoscaling, but GCP doesn't allow you scale down to zero. The minimum in an autoscaling group must be 1 or more.

**Comment by davejrt** [comment]
an update: Using the GCE plugin, we can get a test run in less than 5 minutes from scratch without a pre-warmed image. This test fails but I think this could be the way forward. 

https://buildkite.com/sourcegraph/deploy-sourcegraph-docker/builds/896#bb43ea76-9233-4023-92dc-b1e243053631

**Comment by davejrt** [comment]
Successful run in less than 10 minutes, including the smoke tests

**Comment by slimsag** [comment]
I landed some fixes to my tests in https://github.com/sourcegraph/deploy-sourcegraph-docker/pull/135 which helped improve stability when running locally, but testing on the baremetal agents shows still lots of network issues: https://github.com/sourcegraph/deploy-sourcegraph-docker/pull/136

I plan to next try and SSH directly into one of the buildkite agents and run the tests - there has to be some explanation for why the network here is so unstable?

**Comment by slimsag** [comment]
I'm not opposed to the GCP plugin approach, but it would be really nice to have the option to run the vagrant files locally on our dev machines too

**Comment by pecigonzalo** [comment]
> there has to be some explanation for why the network here is so unstable?
I agree it would be great to understand the "Why"

> it would be really nice to have the option to run the vagrant files locally on our dev machines too

I think there are 2 options, either have 2 providers and select which one to use with `vagrant up --provider=X` or local dev machines also use a remote VM to run, which in some cases might be preferred.

**Comment by davejrt** [comment]
I previously implemented [your gist](https://gist.github.com/pecigonzalo/c86f9e9b5de5464afbfe370f4f9bae27) @pecigonzalo but it didn't yield the results I was hoping for. I searched pretty extensively for vbox settings that might help as well but came up with nothing. The libvirt/kvm providers are not officially supported and old so those don't seem like a viable right route either.

**Comment by davejrt** [comment]
one comment regarding being up to use this locally, I'd say it would be possible to use the yaml format I have with different configs and select the provider as @pecigonzalo mentioned. Not to say we shouldn't get this working on a gcp vm though, but at the very least it would be nice to have the flexibility in my opinion

**Comment by davejrt** [comment]
see sourcegraph/deploy-sourcegraph-docker#140 for some more info on where this is currently at

TLDR; tests run now, however the more cpu we use...the slower they run. 8vcpu's on the virtualbox provider seems to be the sweet spot, regardless of how big the host instance is.

**Comment by davejrt** [comment]
Closing as ultimately it seems that nested-virt with vbox is generally unstable. We have moved forward with solely cloud based testing using `vagrant-google`
